stratosphere/ncm/source/ncm_content_meta_database.cpp

```diff
@@ -0,0 +1,584 @@
+                        break;
+                    default:
+                        /* The first u64 of all non-application extended headers is the application program id. */
+                        application_id = *GetValueExtendedHeader<ProgramId>(value);
+                }
+
+                /* Write the entry to the output buffer. */
+                if (entries_written < out_keys.GetSize()) {
+                    ApplicationContentMetaKey* out_app_key = &out_keys[entries_written];
+                    out_app_key->application_program_id = application_id;
+                    out_app_key->key = key;
+                    entries_written++;
+                }
+
+                entries_total++;
+            }
+        }
+
+        out_entries_total.SetValue(entries_total);
+        out_entries_written.SetValue(entries_written);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::Has(sf::Out<bool> out, ContentMetaKey key) {
+        R_TRY(this->EnsureEnabled());
+
+        if (this->kvs->GetCount() == 0) {
+            out.SetValue(false);
+            return ResultSuccess();
+        }
+
+        bool has = false;
+        const auto it = this->kvs->lower_bound(key);
+        if (it != this->kvs->end()) {
+            has = it->GetKey() == key;
+        }
+
+        out.SetValue(has);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::HasAll(sf::Out<bool> out, const sf::InArray<ContentMetaKey> &keys) {
+        R_TRY(this->EnsureEnabled());
+
+        bool has = true;
+        for (size_t i = 0; i < keys.GetSize() && has; i++) {
+            R_TRY(this->Has(&has, keys[i]));
+        }
+
+        out.SetValue(has);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::GetSize(sf::Out<u64> out_size, ContentMetaKey key) {
+        R_TRY(this->EnsureEnabled());
+        R_UNLESS(this->kvs->GetCount() != 0, ncm::ResultContentMetaNotFound());
+
+        const auto it = this->kvs->lower_bound(key);
+        R_UNLESS(it != this->kvs->end(), ncm::ResultContentMetaNotFound());
+        R_UNLESS(it->GetKey() == key, ncm::ResultContentMetaNotFound());
+
+        out_size.SetValue(it->GetValueSize());
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::GetRequiredSystemVersion(sf::Out<u32> out_version, ContentMetaKey key) {
+        R_TRY(this->EnsureEnabled());
+        R_UNLESS(key.type == ContentMetaType::Application || key.type == ContentMetaType::Patch, ncm::ResultInvalidContentMetaKey());
+
+        const void* value = nullptr;
+        size_t value_size = 0;
+        R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+
+        /* Required system version is at the same offset for the patch and application extended header.
+           We use the application header for convenience. */
+        const auto ext_header = GetValueExtendedHeader<ApplicationMetaExtendedHeader>(value);
```

Comment:
I would prefer that you have an if() { } else {} anyway, to make code less fragile to type definitions changing.

If you want to to it this way, add a static_assert that the relevant OFFSETOF is identical.